Convert to circle

LinuxBozo opened the pull request:
Conversion from Travis to Circle.

Makes use of https://github.com/tpoechtrager/osxcross to do cross compilation for OSX since Mac support on CircleCI is spotty on v1, and not supported on v2 configurations.

LinuxBozo committed:
Switch over to CircleCI setup

LinuxBozo committed:
Enable cross compiling for OSX

cmc333333 commented on .circleci/build-osx-cross.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
```
Do we want a `set -e`/similar here?

LinuxBozo replied:
Good call.

cmc333333 commented on .circleci/build-osx-cross.sh:
```diff
@@ -0,0 +1,6 @@
+cd /tmp
+git clone https://github.com/tpoechtrager/osxcross.git
+cd osxcross
+curl -#L https://github.com/phracker/MacOSX-SDKs/releases/download/10.13/MacOSX10.11.sdk.tar.xz -o tarballs/MacOSX10.11.sdk.tar.xz
```
The `-#L` confuses me -- do we not want redirects?

LinuxBozo replied:
`-#L` is two options to curl.. progress bar, and follow redirects.  It doesn't actually comment as seen in the syntax highlighting.

LinuxBozo replied:
Will use the long form option instead to clear any confusion in the future.

cmc333333 commented on .circleci/config.yml:
```diff
@@ -0,0 +1,56 @@
+      - run: sudo apt-get install -y pkg-config cmake shellcheck clang llvm-dev libxml2-dev uuid-dev libssl-dev file bash patch make gcc g++ zlib1g-dev libmpc-dev libmpfr-dev libgmp-dev tar xz-utils bzip2 gzip sed cpio
+      - run: shellcheck build.sh
+      - run: ./.circleci/build-osx-cross.sh
+      - run: export CROSS_DIR=/tmp/osxcross/target && export PATH=$CROSS_DIR/bin:$PATH && export SDKROOT="$CROSS_DIR/SDK/MacOSX10.11.sdk" && export OSTYPE=darwin15 && ./build.sh
```
Small detail, but would this be cleaner (and work) as either
```
CROSS_DIR=/tmp/osxcross/target PATH=$CROSS_DIR/bin:$PATH SDKROOT="$CROSS_DIR/SDK/MacOSX10.11.sdk" OSTYPE=darwin15 ./build.sh
```
(i.e. no exports)

or
```yaml
run:
    environment:
        CROSS_DIR: /tmp/osxcross/target
        PATH: "$CROSS_DIR/bin:$PATH" 
        SDKROOT: "$CROSS_DIR/SDK/MacOSX10.11.sdk" 
        OSTYPE: darwin15
    command: ./build.sh
```
?
The latter is easier to read, imho.

LinuxBozo replied:
Good call, I like the latter as well.

LinuxBozo replied:
And unfortunately, doesn't work for interpolation like we're doing for vars reliant on CROSS_DIR: https://circleci.com/docs/2.0/env-vars/#interpolating-environment-variables-to-set-other-environment-variables

😞

cmc333333 reviewed (approved):
Neat! I skimmed over the code and it fits a shape I'd expect (though I'm not familiar with the domain). I've ran the compiled artifacts for both OSX and Linux and they seemed to execute just fine.

LinuxBozo committed:
Address review comments

LinuxBozo committed:
Use /bin/bash explicitly

LinuxBozo committed:
Back out environment variables. Doesn't support interpolation:
https://circleci.com/docs/2.0/env-vars/#interpolating-environment-variables-to-set-other-environment-variables
